core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/UpscaleFloatingCodec.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.datastore.page.encoding;
+
+import org.apache.carbondata.core.datastore.compression.Compressor;
+import org.apache.carbondata.core.datastore.page.ColumnPage;
+import org.apache.carbondata.core.datastore.page.LazyColumnPage;
+import org.apache.carbondata.core.datastore.page.PrimitiveCodec;
+import org.apache.carbondata.core.datastore.page.statistics.ColumnPageStatsVO;
+import org.apache.carbondata.core.metadata.datatype.DataType;
+
+/**
+ * Codec for floating point (float, double) data type page.
+ * This codec will upscale the diff from page max value to integer value,
+ * and do type casting to make storage minimum.
+ */
+public class UpscaleFloatingCodec extends AdaptiveCompressionCodec {
+
+  private ColumnPage encodedPage;
+  private double factor;
+
+  public static ColumnPageCodec newInstance(DataType srcDataType, DataType targetDataType,
+      ColumnPageStatsVO stats, Compressor compressor) {
+    return new UpscaleFloatingCodec(srcDataType, targetDataType, stats, compressor);
+  }
+
+  private UpscaleFloatingCodec(DataType srcDataType, DataType targetDataType,
+      ColumnPageStatsVO stats, Compressor compressor) {
+    super(srcDataType, targetDataType, stats, compressor);
+    this.factor = Math.pow(10, stats.getDecimal());
+  }
+
+  @Override public String getName() {
+    return "UpscaleFloatingCodec";
+  }
+
+  @Override
+  public byte[] encode(ColumnPage input) {
+    encodedPage = ColumnPage.newPage(targetDataType, input.getPageSize());
+    input.encode(codec);
+    return encodedPage.compress(compressor);
+  }
+
+
+  @Override
+  public ColumnPage decode(byte[] input, int offset, int length) {
+    ColumnPage page = ColumnPage.decompress(compressor, targetDataType, input, offset, length);
+    return LazyColumnPage.newPage(page, codec);
+  }
+
+  private PrimitiveCodec codec = new PrimitiveCodec() {
```

Comment:
I thought about it earlier, but in that case the log message will not be as good as now, like "internal error: " + debugInfo().